File: backend/pkg/api/article/handlers.go
package article

import (
	"backend/pkg/api/middleware"
	"backend/pkg/api/response"
	"backend/pkg/api/validation"
	"backend/pkg/core"
	coreArticle "backend/pkg/core/article"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

// GenerateArticle handles POST /blog/generate
// @Summary Generate article with AI
// @Description Generate a new blog article using AI based on the provided prompt
// @Tags articles
// @Accept json
// @Produce json
// @Param request body object{prompt=string,title=string,publish=boolean} true "Generation request"
// @Success 200 {object} response.SuccessResponse{data=dto.ArticleResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /blog/generate [post]
func GenerateArticle(c *fiber.Ctx) error {
	var req struct {
		Prompt  string `json:"prompt"`
		Title   string `json:"title"`
		Publish bool   `json:"publish"`
	}
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}

	userID, err := middleware.GetUserID(c)
	if err != nil {
		return response.Error(c, core.UnauthorizedError("User not authenticated"))
	}

	article, err := coreArticle.GenerateArticle(c.Context(), req.Prompt, req.Title, userID, req.Publish)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, article)
}

// UpdateArticle handles POST /blog/articles/:slug/update
// @Summary Update article draft
// @Description Update the draft content of an existing article
// @Tags articles
// @Accept json
// @Produce json
// @Param slug path string true "Article slug"
// @Param request body dto.UpdateArticleRequest true "Update request"
// @Success 200 {object} response.SuccessResponse{data=dto.ArticleResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /blog/articles/{slug}/update [post]
func UpdateArticle(c *fiber.Ctx) error {
	slug := c.Params("slug")
	if slug == "" {
		return response.Error(c, core.InvalidInputError("Article slug is required"))
	}

	articleID, err := coreArticle.GetIDBySlug(c.Context(), slug)
	if err != nil {
		return response.Error(c, err)
	}

	var req coreArticle.UpdateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}

	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	article, err := coreArticle.Update(c.Context(), articleID, req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, article)
}

// CreateArticle handles POST /blog/articles
// @Summary Create new article
// @Description Create a new blog article
// @Tags articles
// @Accept json
// @Produce json
// @Param request body dto.CreateArticleRequest true "Create request"
// @Success 201 {object} response.SuccessResponse{data=dto.ArticleResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /blog/articles [post]
func CreateArticle(c *fiber.Ctx) error {
	var req coreArticle.CreateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}

	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	article, err := coreArticle.Create(c.Context(), req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Created(c, article)
}

// UpdateArticleWithContext handles PUT /blog/:id/update
// @Summary Update article with AI context
// @Description Update article using AI-powered context-aware editing
// @Tags articles
// @Accept json
// @Produce json
// @Param id path string true "Article ID"
// @Success 200 {object} response.SuccessResponse{data=dto.ArticleResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /blog/{id}/update [put]
func UpdateArticleWithContext(c *fiber.Ctx) error {
	articleIDStr := c.Params("id")
	articleID, err := uuid.Parse(articleIDStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid article ID"))
	}

	article, err := coreArticle.UpdateWithContext(c.Context(), articleID)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, article)
}

// GetArticles handles GET /blog/articles
// @Summary List articles
// @Description Get a paginated list of articles with optional filtering
// @Tags articles
// @Accept json
// @Produce json
// @Param page query int false "Page number" default(1)
// @Param articlesPerPage query int false "Articles per page" default(6)
// @Param tag query string false "Filter by tag"
// @Param status query string false "Filter by status (published/draft)" default(published)
// @Param sortBy query string false "Sort field"
// @Param sortOrder query string false "Sort order (asc/desc)"
// @Success 200 {object} response.SuccessResponse{data=dto.ArticleListResponse}
// @Failure 500 {object} response.SuccessResponse
// @Router /blog/articles [get]
func GetArticles(c *fiber.Ctx) error {
	page := c.QueryInt("page", 1)
	tag := c.Query("tag", "")
	status := c.Query("status", "published")
	articlesPerPage := c.QueryInt("articlesPerPage", 6)
	sortBy := c.Query("sortBy", "")
	sortOrder := c.Query("sortOrder", "")

	articles, err := coreArticle.List(c.Context(), page, tag, status, articlesPerPage, sortBy, sortOrder)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, articles)
}

// SearchArticles handles GET /blog/articles/search
// @Summary Search articles
// @Description Search articles by query string
// @Tags articles
// @Accept json
// @Produce json
// @Param query query string true "Search query"
// @Param page query int false "Page number" default(1)
// @Param tag query string false "Filter by tag"
// @Param status query string false "Filter by status" default(published)
// @Param sortBy query string false "Sort field"
// @Param sortOrder query string false "Sort order"
// @Success 200 {object} response.SuccessResponse{data=dto.ArticleListResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Router /blog/articles/search [get]
func SearchArticles(c *fiber.Ctx) error {
	query := c.Query("query")
	if query == "" {
		return response.Error(c, core.InvalidInputError("Query parameter is required"))
	}

	page := c.QueryInt("page", 1)
	tag := c.Query("tag", "")
	status := c.Query("status", "published")
	sortBy := c.Query("sortBy", "")
	sortOrder := c.Query("sortOrder", "")

	articles, err := coreArticle.Search(c.Context(), query, page, tag, status, sortBy, sortOrder)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, articles)
}

// GetPopularTags handles GET /blog/tags/popular
// @Summary Get popular tags
// @Description Get a list of most used tags
// @Tags articles
// @Accept json
// @Produce json
// @Success 200 {object} response.SuccessResponse{data=object{tags=[]string}}
// @Failure 500 {object} response.SuccessResponse
// @Router /blog/tags/popular [get]
func GetPopularTags(c *fiber.Ctx) error {
	tags, err := coreArticle.GetPopularTags(c.Context())
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"tags": tags})
}

// GetArticleData handles GET /blog/articles/:slug
// @Summary Get article by slug
// @Description Get a single article by its slug
// @Tags articles
// @Accept json
// @Produce json
// @Param slug path string true "Article slug"
// @Success 200 {object} response.SuccessResponse{data=dto.ArticleResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Router /blog/articles/{slug} [get]
func GetArticleData(c *fiber.Ctx) error {
	slug := c.Params("slug")
	if slug == "" {
		return response.Error(c, core.InvalidInputError("Slug is required"))
	}

	data, err := coreArticle.GetBySlug(c.Context(), slug)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, data)
}

// GetRecommendedArticles handles GET /blog/articles/:id/recommended
// @Summary Get recommended articles
// @Description Get articles recommended based on the current article
// @Tags articles
// @Accept json
// @Produce json
// @Param id path string true "Article ID"
// @Success 200 {object} response.SuccessResponse{data=[]dto.ArticleResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Router /blog/articles/{id}/recommended [get]
func GetRecommendedArticles(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid article ID"))
	}

	articles, err := coreArticle.GetRecommended(c.Context(), id)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, articles)
}

// DeleteArticle handles DELETE /blog/articles/:id
// @Summary Delete article
// @Description Permanently delete an article
// @Tags articles
// @Accept json
// @Produce json
// @Param id path string true "Article ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /blog/articles/{id} [delete]
func DeleteArticle(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid article ID"))
	}

	if err := coreArticle.Delete(c.Context(), id); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true})
}

// PublishArticle handles POST /blog/articles/:slug/publish
// @Summary Publish article
// @Description Publish an article's draft to make it publicly visible
// @Tags articles
// @Accept json
// @Produce json
// @Param slug path string true "Article slug"
// @Success 200 {object} response.SuccessResponse{data=dto.ArticleResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /blog/articles/{slug}/publish [post]
func PublishArticle(c *fiber.Ctx) error {
	slug := c.Params("slug")
	if slug == "" {
		return response.Error(c, core.InvalidInputError("Article slug is required"))
	}

	articleID, err := coreArticle.GetIDBySlug(c.Context(), slug)
	if err != nil {
		return response.Error(c, err)
	}

	article, err := coreArticle.Publish(c.Context(), articleID)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, article)
}

// UnpublishArticle handles POST /blog/articles/:slug/unpublish
// @Summary Unpublish article
// @Description Unpublish an article, making it no longer publicly visible
// @Tags articles
// @Accept json
// @Produce json
// @Param slug path string true "Article slug"
// @Success 200 {object} response.SuccessResponse{data=dto.ArticleResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /blog/articles/{slug}/unpublish [post]
func UnpublishArticle(c *fiber.Ctx) error {
	slug := c.Params("slug")
	if slug == "" {
		return response.Error(c, core.InvalidInputError("Article slug is required"))
	}

	articleID, err := coreArticle.GetIDBySlug(c.Context(), slug)
	if err != nil {
		return response.Error(c, err)
	}

	article, err := coreArticle.Unpublish(c.Context(), articleID)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, article)
}

// ListVersions handles GET /blog/articles/:slug/versions
// @Summary List article versions
// @Description Get all versions of an article
// @Tags articles
// @Accept json
// @Produce json
// @Param slug path string true "Article slug"
// @Success 200 {object} response.SuccessResponse{data=dto.ArticleVersionListResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /blog/articles/{slug}/versions [get]
func ListVersions(c *fiber.Ctx) error {
	slug := c.Params("slug")
	if slug == "" {
		return response.Error(c, core.InvalidInputError("Article slug is required"))
	}

	articleID, err := coreArticle.GetIDBySlug(c.Context(), slug)
	if err != nil {
		return response.Error(c, err)
	}

	versions, err := coreArticle.ListVersions(c.Context(), articleID)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, versions)
}

// GetVersion handles GET /blog/articles/versions/:versionId
// @Summary Get article version
// @Description Get a specific version of an article
// @Tags articles
// @Accept json
// @Produce json
// @Param versionId path string true "Version ID"
// @Success 200 {object} response.SuccessResponse{data=dto.ArticleVersionResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /blog/articles/versions/{versionId} [get]
func GetVersion(c *fiber.Ctx) error {
	versionIDStr := c.Params("versionId")
	versionID, err := uuid.Parse(versionIDStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid version ID"))
	}

	version, err := coreArticle.GetVersion(c.Context(), versionID)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, version)
}

// RevertToVersion handles POST /blog/articles/:slug/revert/:versionId
// @Summary Revert article to version
// @Description Revert an article's draft to a previous version
// @Tags articles
// @Accept json
// @Produce json
// @Param slug path string true "Article slug"
// @Param versionId path string true "Version ID to revert to"
// @Success 200 {object} response.SuccessResponse{data=dto.ArticleResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /blog/articles/{slug}/revert/{versionId} [post]
func RevertToVersion(c *fiber.Ctx) error {
	slug := c.Params("slug")
	if slug == "" {
		return response.Error(c, core.InvalidInputError("Article slug is required"))
	}

	articleID, err := coreArticle.GetIDBySlug(c.Context(), slug)
	if err != nil {
		return response.Error(c, err)
	}

	versionIDStr := c.Params("versionId")
	versionID, err := uuid.Parse(versionIDStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid version ID"))
	}

	article, err := coreArticle.RevertToVersion(c.Context(), articleID, versionID)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, article)
}

File: backend/src/api/article/handlers.rs
use axum::{
    Json,
    body::Bytes,
    extract::{Path, Query, State},
    http::StatusCode,
};
use uuid::Uuid;

use crate::{
    api::{auth::AuthenticatedAccount, request::JsonBody, response::SuccessResponse},
    core::article::{CreateArticle, UpdateArticle},
    error::AppError,
};

use super::{
    dto::{
        ArticleListQuery, ArticleSearchQuery, DeleteArticleResponse, GenerateArticleRequest,
        GenerateArticleResponse, PopularTagsResponse, PublishArticleRequest, timestamp,
    },
    state::{ArticleState, GenerationRequest},
};

type ApiResult<T> = Result<Json<SuccessResponse<T>>, AppError>;

#[utoipa::path(
    post,
    path = "/blog/generate",
    request_body = GenerateArticleRequest,
    responses(
        (status = 200, body = SuccessResponse<GenerateArticleResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "articles",
    operation_id = "generateArticle"
)]
pub async fn generate_article(
    authenticated: AuthenticatedAccount,
    State(state): State<ArticleState>,
    JsonBody(request): JsonBody<GenerateArticleRequest>,
) -> ApiResult<GenerateArticleResponse> {
    let author_id = authenticated.into_inner().into_inner();
    let prompt = request.prompt.trim();
    let title = request.title.trim();
    if prompt.is_empty() && title.is_empty() {
        return Err(AppError::InvalidInput(
            "either title or prompt is required".to_owned(),
        ));
    }
    let draft_title = if title.is_empty() {
        "Untitled Article"
    } else {
        title
    };
    let message = match (prompt.is_empty(), title.is_empty()) {
        (true, false) => format!("Write a complete blog article about: {title}"),
        (false, false) => format!("Title: {title}\n\n{prompt}"),
        _ => prompt.to_owned(),
    };
    let service = state.service()?;
    let article = service.create_draft_shell(draft_title, author_id).await?;
    let request_id = match state
        .generation_queue()?
        .submit(GenerationRequest {
            message,
            article_id: article.id,
        })
        .await
    {
        Ok(request_id) => request_id,
        Err(error) => {
            if let Err(rollback_error) = service.delete(article.id).await {
                tracing::error!(
                    article_id = %article.id,
                    %rollback_error,
                    "failed to roll back draft after generation queue rejection"
                );
            }
            return Err(error);
        }
    };
    Ok(Json(SuccessResponse::new(GenerateArticleResponse {
        article,
        request_id,
    })))
}

#[utoipa::path(
    post,
    path = "/blog/articles",
    request_body = CreateArticle,
    responses(
        (status = 201, body = SuccessResponse<crate::core::article::ArticleListItem>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "articles",
    operation_id = "createArticle"
)]
pub async fn create_article(
    _authenticated: AuthenticatedAccount,
    State(state): State<ArticleState>,
    JsonBody(request): JsonBody<CreateArticle>,
) -> Result<
    (
        StatusCode,
        Json<SuccessResponse<crate::core::article::ArticleListItem>>,
    ),
    AppError,
> {
    let article = state.service()?.create(request).await?;
    Ok((StatusCode::CREATED, Json(SuccessResponse::new(article))))
}

#[utoipa::path(
    post,
    path = "/blog/articles/{slug}/update",
    params(("slug" = String, Path)),
    request_body = UpdateArticle,
    responses(
        (status = 200, body = SuccessResponse<crate::core::article::ArticleListItem>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "articles",
    operation_id = "updateArticle"
)]
pub async fn update_article(
    _authenticated: AuthenticatedAccount,
    State(state): State<ArticleState>,
    Path(slug): Path<String>,
    JsonBody(request): JsonBody<UpdateArticle>,
) -> ApiResult<crate::core::article::ArticleListItem> {
    if slug.is_empty() {
        return Err(AppError::InvalidInput(
            "article slug is required".to_owned(),
        ));
    }
    let service = state.service()?;
    let article_id = service.get_id_by_slug(&slug).await?;
    Ok(Json(SuccessResponse::new(
        service.update(article_id, request).await?,
    )))
}

#[utoipa::path(
    put,
    path = "/blog/{id}/update",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<crate::core::article::Article>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "articles",
    operation_id = "updateArticleWithContext"
)]
pub async fn update_article_with_context(
    _authenticated: AuthenticatedAccount,
    State(state): State<ArticleState>,
    Path(id): Path<Uuid>,
) -> ApiResult<crate::core::article::Article> {
    Ok(Json(SuccessResponse::new(
        state.service()?.update_with_context(id).await?,
    )))
}

#[utoipa::path(
    get,
    path = "/blog/articles",
    params(ArticleListQuery),
    responses((status = 200, body = SuccessResponse<crate::core::article::ArticleListResponse>)),
    tag = "articles",
    operation_id = "getArticles"
)]
pub async fn get_articles(
    State(state): State<ArticleState>,
    Query(query): Query<ArticleListQuery>,
) -> ApiResult<crate::core::article::ArticleListResponse> {
    Ok(Json(SuccessResponse::new(
        state
            .service()?
            .list(
                query.page.unwrap_or(1),
                query.tag.as_deref().unwrap_or_default(),
                query.status.as_deref().unwrap_or("published"),
                query.articles_per_page.unwrap_or(6),
                query.sort_by.as_deref().unwrap_or_default(),
                query.sort_order.as_deref().unwrap_or_default(),
            )
            .await?,
    )))
}

#[utoipa::path(
    get,
    path = "/blog/articles/search",
    params(ArticleSearchQuery),
    responses(
        (status = 200, body = SuccessResponse<crate::core::article::ArticleListResponse>),
        (status = 400, body = crate::error::ErrorEnvelope)
    ),
    tag = "articles",
    operation_id = "searchArticles"
)]
pub async fn search_articles(
    State(state): State<ArticleState>,
    Query(query): Query<ArticleSearchQuery>,
) -> ApiResult<crate::core::article::ArticleListResponse> {
    if query.query.is_empty() {
        return Err(AppError::InvalidInput(
            "query parameter is required".to_owned(),
        ));
    }
    // The Go service accepted these compatibility parameters but did not apply
    // them to repository search options.
    let _compatibility_parameters = (&query.tag, &query.sort_by, &query.sort_order);
    Ok(Json(SuccessResponse::new(
        state
            .service()?
            .search(
                &query.query,
                query.page.unwrap_or(1),
                query.status.as_deref().unwrap_or("published"),
            )
            .await?,
    )))
}

#[utoipa::path(
    get,
    path = "/blog/tags/popular",
    responses((status = 200, body = SuccessResponse<PopularTagsResponse>)),
    tag = "articles",
    operation_id = "getPopularTags"
)]
pub async fn get_popular_tags(State(state): State<ArticleState>) -> ApiResult<PopularTagsResponse> {
    Ok(Json(SuccessResponse::new(PopularTagsResponse {
        tags: state.service()?.get_popular_tags().await?,
    })))
}

#[utoipa::path(
    get,
    path = "/blog/articles/{slug}",
    params(("slug" = String, Path)),
    responses(
        (status = 200, body = SuccessResponse<crate::core::article::ArticleListItem>),
        (status = 404, body = crate::error::ErrorEnvelope)
    ),
    tag = "articles",
    operation_id = "getArticleData"
)]
pub async fn get_article_data(
    State(state): State<ArticleState>,
    Path(slug): Path<String>,
) -> ApiResult<crate::core::article::ArticleListItem> {
    if slug.is_empty() {
        return Err(AppError::InvalidInput("slug is required".to_owned()));
    }
    Ok(Json(SuccessResponse::new(
        state.service()?.get_by_slug(&slug).await?,
    )))
}

#[utoipa::path(
    get,
    path = "/blog/articles/{id}/recommended",
    params(("id" = Uuid, Path)),
    responses((status = 200, body = SuccessResponse<Option<Vec<crate::core::article::RecommendedArticle>>>)),
    tag = "articles",
    operation_id = "getRecommendedArticles"
)]
pub async fn get_recommended_articles(
    State(state): State<ArticleState>,
    Path(id): Path<Uuid>,
) -> ApiResult<Option<Vec<crate::core::article::RecommendedArticle>>> {
    let articles = state.service()?.get_recommended(id).await?;
    Ok(Json(SuccessResponse::new(
        (!articles.is_empty()).then_some(articles),
    )))
}

#[utoipa::path(
    delete,
    path = "/blog/articles/{slug}",
    params(("slug" = Uuid, Path, description = "Article ID")),
    responses(
        (status = 200, body = SuccessResponse<DeleteArticleResponse>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "articles",
    operation_id = "deleteArticle"
)]
pub async fn delete_article(
    _authenticated: AuthenticatedAccount,
    State(state): State<ArticleState>,
    Path(id): Path<Uuid>,
) -> ApiResult<DeleteArticleResponse> {
    state.service()?.delete(id).await?;
    Ok(Json(SuccessResponse::new(DeleteArticleResponse {
        success: true,
    })))
}

#[utoipa::path(
    post,
    path = "/blog/articles/{slug}/publish",
    params(("slug" = String, Path)),
    request_body = Option<PublishArticleRequest>,
    responses((status = 200, body = SuccessResponse<crate::core::article::ArticleListItem>)),
    security(("bearerAuth" = [])),
    tag = "articles",
    operation_id = "publishArticle"
)]
pub async fn publish_article(
    _authenticated: AuthenticatedAccount,
    State(state): State<ArticleState>,
    Path(slug): Path<String>,
    body: Bytes,
) -> ApiResult<crate::core::article::ArticleListItem> {
    let service = state.service()?;
    let article_id = service.get_id_by_slug(&slug).await?;
    let request = if body.is_empty() {
        PublishArticleRequest::default()
    } else {
        serde_json::from_slice(&body).unwrap_or_default()
    };
    let published_at = match request.published_at {
        Some(value) => Some(
            timestamp(value)
                .ok_or_else(|| AppError::InvalidInput("published_at is out of range".to_owned()))?,
        ),
        None => None,
    };
    Ok(Json(SuccessResponse::new(
        service.publish(article_id, published_at).await?,
    )))
}

#[utoipa::path(
    post,
    path = "/blog/articles/{slug}/unpublish",
    params(("slug" = String, Path)),
    responses((status = 200, body = SuccessResponse<crate::core::article::ArticleListItem>)),
    security(("bearerAuth" = [])),
    tag = "articles",
    operation_id = "unpublishArticle"
)]
pub async fn unpublish_article(
    _authenticated: AuthenticatedAccount,
    State(state): State<ArticleState>,
    Path(slug): Path<String>,
) -> ApiResult<crate::core::article::ArticleListItem> {
    let service = state.service()?;
    let article_id = service.get_id_by_slug(&slug).await?;
    Ok(Json(SuccessResponse::new(
        service.unpublish(article_id).await?,
    )))
}

#[utoipa::path(
    get,
    path = "/blog/articles/{slug}/versions",
    params(("slug" = String, Path)),
    responses((status = 200, body = SuccessResponse<crate::core::article::ArticleVersionListResponse>)),
    security(("bearerAuth" = [])),
    tag = "articles",
    operation_id = "listArticleVersions"
)]
pub async fn list_versions(
    _authenticated: AuthenticatedAccount,
    State(state): State<ArticleState>,
    Path(slug): Path<String>,
) -> ApiResult<crate::core::article::ArticleVersionListResponse> {
    let service = state.service()?;
    let article_id = service.get_id_by_slug(&slug).await?;
    Ok(Json(SuccessResponse::new(
        service.list_versions(article_id).await?,
    )))
}

#[utoipa::path(
    get,
    path = "/blog/articles/versions/{versionId}",
    params(("versionId" = Uuid, Path)),
    responses((status = 200, body = SuccessResponse<crate::core::article::ArticleVersionResponse>)),
    security(("bearerAuth" = [])),
    tag = "articles",
    operation_id = "getArticleVersion"
)]
pub async fn get_version(
    _authenticated: AuthenticatedAccount,
    State(state): State<ArticleState>,
    Path(version_id): Path<Uuid>,
) -> ApiResult<crate::core::article::ArticleVersionResponse> {
    Ok(Json(SuccessResponse::new(
        state.service()?.get_version(version_id).await?,
    )))
}

#[utoipa::path(
    post,
    path = "/blog/articles/{slug}/revert/{versionId}",
    params(("slug" = String, Path), ("versionId" = Uuid, Path)),
    responses((status = 200, body = SuccessResponse<crate::core::article::ArticleListItem>)),
    security(("bearerAuth" = [])),
    tag = "articles",
    operation_id = "revertArticleToVersion"
)]
pub async fn revert_to_version(
    _authenticated: AuthenticatedAccount,
    State(state): State<ArticleState>,
    Path((slug, version_id)): Path<(String, Uuid)>,
) -> ApiResult<crate::core::article::ArticleListItem> {
    let service = state.service()?;
    let article_id = service.get_id_by_slug(&slug).await?;
    Ok(Json(SuccessResponse::new(
        service.revert_to_version(article_id, version_id).await?,
    )))
}
